lib/reverse_sentence.rb

```diff
@@ -1,6 +1,80 @@
 # A method to reverse the words in a sentence, in place.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n): there are two loops, first loop is O(n/2) and the second loop is O(n).
+#                        So, O(n/2) + O(n) --> O(n)
+# Space complexity: O(1): In the first loop on line 16, it reverses the string in place.
+#                         In the second loop on line 24, the helper method "reverse_a_word" reverses each word in place,
+#                         and the reversed word replaces the original String[x..y], it's in place replacement.
+#                         Plus extra variables, overall, the space complexity is a constant.
+
 def reverse_sentence(my_sentence)
-  raise NotImplementedError
+  return "" if my_sentence.nil? || my_sentence.empty?
+  if my_sentence.split.length == 1
+    return my_sentence
+  else
+    # Step 1: Reverse the whole String to keep all whitespaces
+    idx = 0
+    while idx < (my_sentence.length / 2)
+      my_sentence[idx], my_sentence[my_sentence.length - idx - 1] = my_sentence[my_sentence.length - idx - 1], my_sentence[idx]
+      idx += 1
+    end
+
+    # Step 2: Reverse each word in the String back to its original word
+    start_idx = 0
+    end_idx = 0
+    while end_idx < (my_sentence.length - 1)
+      # call the helper method reverse_a_word to reverse the words
+      if my_sentence[end_idx].match?(/\s/)
+        # when the first element in the string is not a whitespace
+        if end_idx > 0
+          my_sentence[start_idx..(end_idx - 1)] = reverse_a_word(my_sentence[start_idx..(end_idx - 1)])
+        # when the first element in the string is a whitespace
+        else
+          my_sentence[start_idx..(end_idx)] = reverse_a_word(my_sentence[start_idx..(end_idx)])
+        end
+        start_idx = end_idx + 1
+      end
+      end_idx += 1
+    end
+
+    # Step 3: Handle the last word
+    my_sentence[(start_idx)..(my_sentence.length)] = reverse_a_word(my_sentence[(start_idx)..(my_sentence.length)])
```

Comment:
Just note that this creates space for a word, however if you assume the words are pretty short, it's negligible.

Reply:
I was thinking about using gsub! for in place replacement, but I was also wondering, how does gsub! work?  If gsub! didn't use extra space for in place replacement, I could do that too? But I just couldn't figure it out...orz